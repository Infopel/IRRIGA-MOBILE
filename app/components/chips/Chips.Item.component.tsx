import { MaterialIcons } from "@expo/vector-icons"
import { HStack, Icon, IconButton, Pressable, Text } from "native-base"
import * as React from "react"
import logger from "utils/logger"
interface ChipItemProps {
  children: string
  onDeletePressed?: () => void
  onPress?: () => void
  size?: "sm" | "md"
  status?: "selected" | "action" | "none" | "edit"
}

const ChipItem: React.FunctionComponent<ChipItemProps> = ({
  children,
  onPress,
  status,
  size = "md",
  onDeletePressed,
}) => {
  const _size = {
    md: {
      pressable: {
        py: 2,
        mr: 1,
        pr: 1,
        mb: 1,
      },
      text: { fontSize: "sm" },
      hStack: {
        pr: 4,
      },
    },
    sm: {
      pressable: {
        py: 1,
        mr: 1,
        pr: 1,
        mb: 1,
      },
      text: {
        fontSize: "xs",
      },
      hStack: {
        pr: 2,
      },
    },
  }

  const currentSize = _size[size]

  const isSelected = status === "selected"
  const isAction = status === "action" || status === "edit"

  return (
    <Pressable
      borderRadius={"3xl"}
      {...currentSize.pressable}
      bg={isSelected ? "primary.400" : isAction ? "primary.400" : "tertiary.100"}
      _pressed={{
        bg: "blueGray.400",
      }}
      accessibilityLabel={isAction ? `${status}-chip` : `chip-item-${children}`}
      accessibilityState={{ selected: isSelected }}
      disabled={!onPress}
      onPress={onPress}
    >
      {({ isFocused, isHovered, isPressed }) => (
        <HStack alignItems={"center"} pl={isAction ? 2 : 4} space={2} {...currentSize.hStack}>
          {isAction && (<IconButton
            m='0' p='0' variant={'ghost'}
            accessibilityLabel={`action-${children}`}
            onPress={onPress}
            color={"white"}
            icon={<Icon
              as={MaterialIcons}
              name={status !== "action" ? "edit" : "add-circle"}
              size={"sm"}
            />}
          />
          )}
          <Text
            {...currentSize.text}
            color={isPressed || isSelected || isAction ? "white" : "primary.900"}
          >
            {children}
          </Text>
          {onDeletePressed && (<IconButton m='0' p='0'
            accessibilityLabel={`delete-${children}`}
            onPress={onDeletePressed}
            color={isPressed || isSelected || isAction ? "white" : "primary.900"}
            size={"sm"}
            variant={'ghost'}
            icon={<Icon
              as={MaterialIcons}
              name="cancel"
            />}
          />
          )}
        </HStack>
      )}
    </Pressable>
  )
}

export default ChipItem
