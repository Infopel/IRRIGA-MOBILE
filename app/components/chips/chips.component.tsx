import { HStack, IBoxProps } from "native-base"
import * as React from "react"
import { StyleProp, ViewStyle } from "react-native"
import ChipItem from "./Chips.Item.component"

export type Item = {
  id: string
  name: string
}

interface BaseChipsProps extends IBoxProps {
  /**
   * An optional style override useful for padding & margin.
   */
  items: Item[]
  size?: "sm" | "md"
}

interface ActionChipsProps extends BaseChipsProps {
  onAction?: () => void
  hideActionBtn?: boolean
  onItemClick?: (item: string, action: "delete" | "press") => void
}
interface SelectableChipsProps extends BaseChipsProps {
  selectedItems: string[]
  onSelectItems: (selectedItem: string | null, items: string[]) => void
}
export type ChipsProps = SelectableChipsProps | ActionChipsProps

function isActionChips(args: ChipsProps): args is ActionChipsProps {
  return (args as ActionChipsProps).onAction !== undefined
}
function isSelectableChips(args: ChipsProps): args is SelectableChipsProps {
  return (
    (args as SelectableChipsProps).onSelectItems !== undefined &&
    (args as SelectableChipsProps).selectedItems !== undefined
  )
}

/**
 * Describe your component here
 */
export const Chips = (props: ChipsProps) => {
  //@ts-ignore
  const { items, size, onChangeItems: oc, selectedItems: si, onAction: ao, hideActionBtn: ha, onItemClick: oi, ...rest } = props

  const onChangeItems = isSelectableChips(props) ? props.onSelectItems : undefined
  const selectedItems = isSelectableChips(props) ? props.selectedItems : undefined
  const onAction = isActionChips(props) ? props.onAction : undefined
  const hideActionButton = isActionChips(props) ? props.hideActionBtn : false
  const onItemClick = isActionChips(props) ? props.onItemClick : undefined

  const _isDeletable = onItemClick || onChangeItems
  const _isSelectable = onItemClick || onChangeItems

  const handleOnSelectItem = (itemId: string) => {
    if (!isSelectableChips(props)) {
      onItemClick?.(itemId, "press")
      return
    }

    const _selectedItems = selectedItems ?? []
    onChangeItems?.call(this, itemId, [..._selectedItems, itemId + ""])
  }
  const handleOnDelete = (itemId: string) => {
    if (isSelectableChips(props)) {
      onChangeItems?.call(this, null, selectedItems?.filter((id) => id !== itemId) ?? [])
    } else {
      onItemClick?.(itemId, "delete")
    }
  }

  const visisbleItems =
    items?.map((item) =>
      selectedItems?.find((i) => i === item.id)
        ? { ...item, isSelected: true }
        : { ...item, isSelected: false },
    ) ?? []

  return (
    <HStack {...rest} flexWrap="wrap">
      {visisbleItems.map(({ id, name, isSelected }) => (
        <ChipItem
          key={id}
          size={size}
          status={isSelected ? "selected" : "none"}
          onDeletePressed={!_isDeletable ? undefined : handleOnDelete.bind(this, id + "")}
          onPress={!_isSelectable ? undefined : handleOnSelectItem.bind(this, id + "")}
        >
          {name}
        </ChipItem>
      ))}
      {onAction && !hideActionButton && (
        <ChipItem
          size={size}
          status={visisbleItems.length <= 0 ? "action" : "edit"}
          onPress={onAction}
        >
          {visisbleItems.length === 0 ? "Adicionar" : "Alterar"}
        </ChipItem>
      )}
    </HStack>
  )
}
