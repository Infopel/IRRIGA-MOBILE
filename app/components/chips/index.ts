export * from './Chips.Item.component'
export * from './chips.component'