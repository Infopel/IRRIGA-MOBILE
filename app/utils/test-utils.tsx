import DatabaseProvider from "@nozbe/watermelondb/DatabaseProvider"
import { render } from "@testing-library/react-native"
import React from "react"
import { database } from "storage"
import { NativeBaseProvider as ThemeProvider } from "theme"
export type AllTheProvidersProps = { children: React.ReactNode }
function AllTheProviders({ children }: AllTheProvidersProps) {
  return (
    <DatabaseProvider database={database}>
      <ThemeProvider>{children}</ThemeProvider>
    </DatabaseProvider>
  )
}

type RenderOptions = {
  wrapper?: React.ComponentType<any>;
  createNodeMock?: (element: React.ReactElement) => any;
}

function customRender<T extends any>(ui: React.ReactElement<T>, options?: RenderOptions) {
  return render(ui, { wrapper: AllTheProviders, ...options })
}

// re-export everything
export * from "@testing-library/react-native"
// override render method
export { customRender as render }
