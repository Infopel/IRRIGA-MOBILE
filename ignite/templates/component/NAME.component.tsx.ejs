import * as React from "react"
import { StyleProp, TextStyle, View, ViewStyle } from "react-native"
import { color, typography } from "../../theme"
import { Text } from "../../components"
import {<%= props.pascalCaseName %>Styles as styles} from "./<%= props.kebabCaseName %>.styles"

export interface <%= props.pascalCaseName %>Props {
  /**
   * An optional style override useful for padding & margin.
   */
  style?: StyleProp<ViewStyle>
}

/**
 * Describe your component here
 */
export const <%= props.pascalCaseName %> = function <%= props.pascalCaseName %>(props: <%= props.pascalCaseName %>Props) {
  const { style } = props
  const containerStyle = {
    ...styles.container,
    ...(style && typeof style === "object" ? style : {}),
  }

  return (
    <View style={containerStyle}>
      <Text style={styles.text}>Hello</Text>
    </View>
  )
}
