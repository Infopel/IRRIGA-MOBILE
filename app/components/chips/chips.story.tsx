import { storiesOf } from "@storybook/react-native"
import { NativeBaseProvider, theme } from "native-base"
import * as React from "react"
import { Story, StoryScreen, UseCase } from "../../../storybook/views"
import { Chips } from "./chips.component"

storiesOf("Chips", module)
  .addDecorator((fn) => <StoryScreen>{fn()}</StoryScreen>)
  .add("Style Presets", () => (
    <Story>
      <UseCase text="Simple chips" usage="The primary.">
        <Chips
          items={[
            { id: 1, name: "Sopa" },
            { id: 2, name: "Batata" },
            { id: 3, name: "Lagosta" },
            { id: 4, name: "Lola" },
          ]}
        />
      </UseCase>
      <UseCase text="With add button" usage="The primary.">
        <Chips
          items={[
            { id: 1, name: "Sopa" },
            { id: 2, name: "Batata" },
            { id: 3, name: "Lagosta" },
            { id: 4, name: "Lola" },
          ]}
          onAction={() => {}}
        />
      </UseCase>

      <UseCase text="Removable items" usage="The primary.">
        <Chips
          items={[
            { id: 1, name: "Sopa" },
            { id: 2, name: "Batata" },
            { id: 3, name: "Lagosta" },
            { id: 4, name: "Lola" },
          ]}
          selectedItems={["1","2"]}
          onSelectItems={()=>{}}
          // onAction={() => {}}
        />
      </UseCase>
    </Story>
  ))
