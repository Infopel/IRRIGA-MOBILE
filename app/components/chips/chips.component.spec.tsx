import React from 'react'
import { fireEvent, render } from 'test-utils'
import { Chips, Item } from './chips.component'
describe("Chips", () => {
    const defaultItems = [{ id: "1", "name": "Banana" }, { id: "2", "name": "Cabbage" }, { id: "3", "name": "Onions" }, { id: "4", "name": "Tomato" }] as Item[]
    it("render empty chips", () => {
        const selectedItems = [] as string[]
        const onChangeItems = jest.fn()
        const ___defaultItems = [] as Item[]
        const { toJSON } = render(<Chips items={___defaultItems} onSelectItems={onChangeItems} selectedItems={selectedItems} />)
        expect(toJSON()).toMatchSnapshot()
    })
    it("should render Selectable chips ", () => {
        const selectedItems = [] as string[]
        const onChangeItems = jest.fn()
        const { queryByText, queryAllByLabelText } = render(<Chips items={defaultItems} onSelectItems={onChangeItems} selectedItems={selectedItems} />)

        expect(queryAllByLabelText(/chip-item/i).length).toBe(defaultItems.length)
        for (const item of defaultItems) {
            expect(queryByText(item.name)).toBeTruthy()
        }
    })
    it("should render select 2 chips", () => {
        const selectedItems = ["3", "1"] as string[]
        const onChangeItems = jest.fn()
        const { debug, queryByText, getAllByLabelText, getAllByA11yState } = render(<Chips items={defaultItems} onSelectItems={onChangeItems} selectedItems={selectedItems} />)

        expect(getAllByLabelText(/chip-item/i).length).toBe(defaultItems.length)
        expect(getAllByA11yState({ selected: true }).length).toBe(selectedItems.length)
        for (const item of selectedItems) {

            expect(queryByText(defaultItems.find((x) => x.id === item)?.name as string)).toBeTruthy()
        }
    })
    it("should onSelectedItems be called", () => {
        const selectedItems = [] as string[]
        const onChangeItems = jest.fn()
        const { getByText } = render(<Chips items={defaultItems} onSelectItems={onChangeItems} selectedItems={selectedItems} />)

        const selectedItem = defaultItems[1]
        fireEvent.press(getByText(selectedItem.name))
        expect(onChangeItems).toBeCalledWith(
            selectedItem.id, [selectedItem.id]
        )

    })
    it("render action - filled action chip", () => {
        const onItemClick = jest.fn()
        const onAction = jest.fn()
        const { debug, queryByText, queryAllByLabelText, queryByLabelText } = render(<Chips items={defaultItems} onAction={onAction} onItemClick={onItemClick} />)

        expect(queryAllByLabelText(/chip-item/i).length).toBe(defaultItems.length)
        expect(queryByLabelText('edit-chip')).toBeTruthy()
        for (const item of defaultItems) {
            expect(queryByText(item.name)).toBeTruthy()
        }
    })
    it("render action - empty action chip", () => {
        const onItemClick = jest.fn()
        const onAction = jest.fn()
        const ___defaultItems = [] as Item[]
        const { queryAllByLabelText, queryByLabelText } = render(<Chips items={___defaultItems} onAction={onAction} onItemClick={onItemClick} />)

        expect(queryAllByLabelText(/chip-item/i).length).toBe(0)
        expect(queryByLabelText('action-chip')).toBeTruthy()
    })
    it("render action - action chip", () => {
        const onItemClick = jest.fn()
        const onAction = jest.fn()
        const ___defaultItems = [] as Item[]
        const { queryAllByLabelText, queryByLabelText } = render(<Chips items={___defaultItems} onAction={onAction} onItemClick={onItemClick} />)

        expect(queryAllByLabelText(/chip-item/i).length).toBe(0)
        expect(queryByLabelText('action-chip')).toBeTruthy()
    })
    it("should on action button be called", () => {
        const onItemClick = jest.fn()
        const onAction = jest.fn()
        const ___defaultItems = [] as Item[]
        const { queryAllByLabelText, getByLabelText } = render(<Chips items={___defaultItems} onAction={onAction} onItemClick={onItemClick} />)
        fireEvent.press(getByLabelText('action-chip'))
        expect(onAction).toBeCalled()
    })
    it("should on item press button be called", () => {
        const onItemClick = jest.fn()
        const onAction = jest.fn()
        const { getByText } = render(<Chips items={defaultItems} onAction={onAction} onItemClick={onItemClick} />)

        const selectedItem = defaultItems[1]
        fireEvent.press(getByText(selectedItem.name))
        expect(onItemClick).toBeCalledWith(
            selectedItem.id, 'press'
        )

    })
    it("should on item delete button be called", () => {
        const onItemClick = jest.fn()
        const onAction = jest.fn()
        const { getByLabelText } = render(<Chips items={defaultItems} onAction={onAction} onItemClick={onItemClick} />)

        const selectedItem = defaultItems[1]
        fireEvent.press(getByLabelText("delete-" + selectedItem.name))
        expect(onItemClick).toBeCalledWith(
            selectedItem.id, 'delete'
        )

    })
})