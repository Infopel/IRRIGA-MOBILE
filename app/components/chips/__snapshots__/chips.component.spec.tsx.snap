// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chips render empty chips 1`] = `
<View
  dataSet={Object {}}
  onSelectItems={[MockFunction]}
  style={
    Object {
      "flexDirection": "row",
      "flexWrap": "wrap",
    }
  }
/>
`;
